# Homework-3 

## Use the Titanic Dataset to Practice Data Wrangling Tasks
### Part 1: Import libaries , load the data, inspect the data, then get the name of the variables and create a small markdown table  with variable names and types (e.g numerical or categorical ordinal, categorical nominal etc.)

### Part 2: 
    - Impute missing values for "Embarked" with the mode (most frequent value)
    - Impute missing value for Age using two following strategies
        - A) Simple mean imputation (fill with the overal mean age)
        - B) Group-wise median imputation: for each combination of Sex and Pclass compute the median age. Impute missing age using the median age of passenger's sex and Pclass group

### Part 3: Create a FamilySize variable (FamilySize = SibSp + Parch +1) and compute:
    - Minimum, Maximum, and mean FamilySize
    - How many passengers were traveling alone
    
### Part 4:  Create a filtered dataframe of all passengers:
    - Who are female
    - Who are male 
    - Who are under 16 years old   
    - Compute the survival rate for the above filtered data
    - Compute survival rate by Sex and Pclass (use groupby and mean) and present the result in a markedown table (rows: sex, columns: Pclass) or using print formating.
    - Interpret your result in 2-3 sentences
     
    
### Part 5: Compute the average fare for each Pclass (Fare values are per tickets not per person (there coud be more than 1 person on a ticket) so make sure to account for that. 

# Part 1

In [421]:
import pandas as pd
df = pd.read_csv('titanic.csv')
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 83.7 KB


| Variable | Data type |
|----------|----------|
| PassengerID  | Numerical nominal  |
| Survived  | Categorical nominal   |
| Pclass  |  Categorical nominal  |
| Name  | Categorical nominal   |
| Sex  | Categorical nominal   |
| Age  | Numerical ratio   |
| SibSp  | Numerical ratio   |
| Parch  | Categorical ratio   |
| Ticket  | Categorical nominal   |
| Fare  | Numerical ratio   |
| Cabin  | Categorical nominal   |
| Embarked  | Categorical nominal   |

# Part 2

In [422]:
df['Embarked'] = df['Embarked'].fillna(df['Embarked'].mode())
df['Age'] = df['Age'].fillna(df['Age'].mean())
agesex_median = df.groupby(['Sex', 'Pclass'])['Age'].transform('median')
df['Age'] = df['Age'].fillna(agesex_median)
df.head(15)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.000000,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.000000,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.000000,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.000000,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.000000,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,29.699118,0,0,330877,8.4583,NaN,Q
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.000000,0,0,17463,51.8625,E46,S
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.000000,3,1,349909,21.0750,NaN,S
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.000000,0,2,347742,11.1333,NaN,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.000000,1,0,237736,30.0708,NaN,C


# Part 3

In [423]:
FamilySize = df['SibSp'] + df['Parch'] + 1
FamilySize.describe()

count    891.000000
mean       1.904602
std        1.613459
min        1.000000
25%        1.000000
50%        1.000000
75%        2.000000
max       11.000000
dtype: float64

The mean family size was 1.9 (approximately 2), the minimum family size was 1, and the maximum family size was 11 on the Titanic.

In [424]:
df[FamilySize<=1].count()

PassengerId    537
Survived       537
Pclass         537
Name           537
Sex            537
Age            537
SibSp          537
Parch          537
Ticket         537
Fare           537
Cabin           94
Embarked       535
dtype: int64

There were 537 total people that were alone

# Part 4

In [425]:
female_df = df[df['Sex']=='female']
male_df = df[df['Sex']=='male']
young_df = df[df['Age']<16]
female_df.count()

PassengerId    314
Survived       314
Pclass         314
Name           314
Sex            314
Age            314
SibSp          314
Parch          314
Ticket         314
Fare           314
Cabin           97
Embarked       312
dtype: int64

In [426]:
female_df[female_df['Survived']==1].count()

PassengerId    233
Survived       233
Pclass         233
Name           233
Sex            233
Age            233
SibSp          233
Parch          233
Ticket         233
Fare           233
Cabin           91
Embarked       231
dtype: int64

233 / 314 = 74.20% of females survived.

In [427]:
male_df.count()

PassengerId    577
Survived       577
Pclass         577
Name           577
Sex            577
Age            577
SibSp          577
Parch          577
Ticket         577
Fare           577
Cabin          107
Embarked       577
dtype: int64

In [428]:
male_df[male_df['Survived']==1].count()

PassengerId    109
Survived       109
Pclass         109
Name           109
Sex            109
Age            109
SibSp          109
Parch          109
Ticket         109
Fare           109
Cabin           45
Embarked       109
dtype: int64

109 / 577 = 18.89% of males survived

In [429]:
young_df.count()

PassengerId    83
Survived       83
Pclass         83
Name           83
Sex            83
Age            83
SibSp          83
Parch          83
Ticket         83
Fare           83
Cabin          13
Embarked       83
dtype: int64

In [430]:
young_df[young_df['Survived']==1].count()

PassengerId    49
Survived       49
Pclass         49
Name           49
Sex            49
Age            49
SibSp          49
Parch          49
Ticket         49
Fare           49
Cabin          11
Embarked       49
dtype: int64

49 / 83 = 59.03% of young people survived.

In [431]:
df.groupby(['Sex', 'Pclass'])['Survived'].mean()

Sex     Pclass
female  1         0.968085
        2         0.921053
        3         0.500000
male    1         0.368852
        2         0.157407
        3         0.135447
Name: Survived, dtype: float64

# Survival Rate by percentage, sex and class
| Sex | Passenger Class 1 | Passenger Class 2 | Passenger Class 3 |
|----------|----------|----------|----------|
| Female  | 96.81%  | 92.11%  | 50.00%  |
| Male  | 36.89%  | 15.74%  | 13.54%  |

Regardless of sex, passenger class 1 had the highest survival rate (e.g. compare 37% to 13-15% survival for males). But females had a higher survival rate overrall, with the difference in highest survival rate being 59.92% (female - male; 96.81% - 36.89%). Thus, you are more likely to survive the Titanic disaster as a female in the first passenger class.

# Part 5

In [432]:
tickets = df.groupby('Ticket')['Ticket'].transform('count')
df['Fare_individual'] = df['Fare'] / tickets
df.groupby('Pclass')['Fare_individual'].mean()

Pclass
1    43.650347
2    13.322599
3     8.085857
Name: Fare_individual, dtype: float64